src/sentry/web/frontend/oauth_token.py

```diff
@@ -41,85 +43,103 @@ def error(self, request: HttpRequest, name, reason=None, status=400):
-            refresh_token = request.POST.get("refresh_token")
-            scope = request.POST.get("scope")
-            client_id = request.POST.get("client_id")
-
-            if not refresh_token:
-                return self.error(request, "invalid_request")
-
-            # TODO(dcramer): support scope
-            if scope:
-                return self.error(request, "invalid_request")
-
-            if not client_id:
-                return self.error(request, "invalid_client", "missing client_id")
-
-            try:
-                application = ApiApplication.objects.get(
-                    client_id=client_id, status=ApiApplicationStatus.active
-                )
-            except ApiApplication.DoesNotExist:
-                return self.error(request, "invalid_client", "invalid client_id")
-
-            try:
-                token = ApiToken.objects.get(application=application, refresh_token=refresh_token)
-            except ApiToken.DoesNotExist:
-                return self.error(request, "invalid_grant", "invalid token")
-
-            token.refresh()
+            return self._get_access_tokens(request, client_id)
+        elif grant_type == GrantTypes.REFRESH:
+            return self._get_refresh_token(request, client_id)
         else:
             return self.error(request, "unsupported_grant_type")
 
+    def _get_access_tokens(self, request, client_id):
+        redirect_uri = request.POST.get("redirect_uri")
+        code = request.POST.get("code")
+
+        try:
+            application = ApiApplication.objects.get(
+                client_id=client_id, status=ApiApplicationStatus.active
+            )
+        except ApiApplication.DoesNotExist:
+            return self.error(request, "invalid_client", "invalid client_id")
+
+        try:
+            grant = ApiGrant.objects.get(application=application, code=code)
+        except ApiGrant.DoesNotExist:
+            return self.error(request, "invalid_grant", "invalid grant")
+
+        if grant.is_expired():
+            return self.error(request, "invalid_grant", "grant expired")
+
+        if not redirect_uri:
+            redirect_uri = application.get_default_redirect_uri()
+        elif grant.redirect_uri != redirect_uri:
+            return self.error(request, "invalid_grant", "invalid redirect_uri")
+
+        access_token = ApiToken.from_grant(grant)
+        id_token = self._get_open_id_token(grant, request)
+        return self._process_token_details(access_token, id_token)
+
+    def _get_open_id_token(self, grant, request):
+        if grant.has_scope("openid"):
+            open_id_token = OpenIDToken(
+                request.POST.get("client_id"),
+                grant.user.id,
+                # Encrypt with a random secret until we implement secure shared secrets in prod
+                secrets.token_urlsafe(),
+                nonce=request.POST.get("nonce"),
+            )
+            return open_id_token.get_encrypted_id_token()
```

Comment:
Won't the ID token also need to know which scopes are authorized so that email/profile scope details can be included as required?

Reply:
The ID token actually just exists on its own mostly, its not used to access API's or protected resources at any point. It's meant to be read by the client upon a users authentication, and it just contains information about what happened during a users authentication + confirmation that the authentication was successful ([source](https://oauth.net/id-tokens-vs-access-tokens/#:~:text=An%20ID%20token%20contains%20information,requirement%20of%20an%20ID%20token.)). The fact that we put email in here is just a "nice to have", it's not required in the spec.

The followup PR to this one ([here](https://github.com/getsentry/sentry/pull/52493)) adds the `/userinfo` endpoint which is the other major part of OIDC, that's a specialized protected resource where clients with the `oidc` scope can retrieve more details on the user, and that's accessed with the standard access token.

Reply:
Yeah, ID tokens aren't used to access APIs/protected resources, but they can also contain profile/email data ([example of that](https://auth0.com/blog/id-token-access-token-what-is-the-difference/)). If profile/email data is part of the ID token, client applications can avoid using `/userinfo` most of the time.

Reply:
Yeah good point - Since it's a separable piece of work from this one I implemented it in a new PR [here](https://github.com/getsentry/sentry/pull/52556). If you don't want this PR to go through without that functionality we can merge the ID token details PR into this one first, or we can merge this one first then do that one after - your call!

Reply:
> If you don't want this PR to go through without that functionality we can merge the ID token details PR into this one first, or we can merge this one first then do that one after - your call!

I'm not too fussy. Great to see that our ID tokens will have the user profile data in it so that consumers aren't required to call `/userinfo` as well.